### PR TITLE
feat!: Make dot access ellipsis match field access

### LINKED ISSUE #5231: TS/JS - dot ellipsis dot should also match field access chain (not just method calls chain)
**Describe the bug**
Probably because of lambdas' but the current assignment does not work with arbitrary ellipsis, which is very important to capture values where a value may be assigned. 

**To Reproduce**
Does not work:
https://semgrep.dev/s/90Dy
Works:
https://semgrep.dev/s/r3OJ

**Expected behavior**
Should capture on Arbitrary ellipsis 

**What is the priority of the bug to you?**

- [x] P1: important to fix or quite annoying

**Use case**
TS/React to 🚀 🌔

### PR DESCRIPTION
**Pull request by nmote**
Previously, `foo. ... .bar` would not match `foo.baz.bar`, but would
match `foo.baz().bar`. I found this confusing, and it seems that I am
not alone (#5231).

Test plan:

Automated tests. Before this change, the following examples in the added
test did not match:

```
foo.baz.bar;
foo.a.b.c.bar;
foo.a.b().c.bar;
```

Now, they all do. In addition, there is a new match in the changed Hack
test.

Also ran `semgrep --test --strict --test-ignore-todo --quiet $PWD`  after making the change in https://github.com/returntocorp/semgrep-rules/tree/nmote/dots-field-chaining and it passed (that was previously the command that failed).

Fixes #5231

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by nmote** [comment]
I have two questions for reviewers:
* This is showing a 5.8% average regression in the benchmarks I noticed this because the previous commit I pushed had a 7+% regression, which triggered a failure. Two recent commits on `develop` have a 2-3% regression, so I suspect that this commit does in fact come with a performance penalty, probably in the vicinity of 3-5%. It makes sense, since it's doing more work when a rule contains a dot access ellipsis, but I wanted to call this out to reviewers and make sure that this is acceptable.
* This changes the output of the `check-send-file.rb` test case in the `semgrep-rules` repo. This looks legit to me, but I'm not sure how to proceed. Should we just merge this despite the failure and then update the `semgrep-rules` repo separately?

File: semgrep-core/src/matching/Generic_vs_generic.ml [hunk 1]
```diff
@@ -126,15 +126,21 @@ let rec obj_and_method_calls_of_expr e =
   match e.G.e with
```
Comment (aryx) [inline semgrep-core/src/matching/Generic_vs_generic.ml:126]: maybe rename obj_and_dot_accesses_of_expr?

File: semgrep-core/src/matching/Generic_vs_generic.ml [hunk 2]
```diff
@@ -126,15 +126,21 @@ let rec obj_and_method_calls_of_expr e =
...
   | B.Call ({ e = B.DotAccess (e, tok, fld); _ }, args) ->
       let o, xs = obj_and_method_calls_of_expr e in
-      (o, (fld, tok, args) :: xs)
+      (o, (fld, tok, Some args) :: xs)
+  | B.DotAccess (e, tok, fld) ->
+      let o, xs = obj_and_method_calls_of_expr e in
+      (o, (fld, tok, None) :: xs)
   | _ -> (e, [])
 
 let rec expr_of_obj_and_method_calls (obj, xs) =
```
Comment (aryx) [inline semgrep-core/src/matching/Generic_vs_generic.ml]: maybe rename expr_of_obj_and_dot_accesses?

File: semgrep-core/tests/js/dot_access_ellipsis.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// MATCH:
```
Comment (aryx) [inline semgrep-core/tests/js/dot_access_ellipsis.js]: there's some kinds of convention on how to name those tests (you can see generate_cheatsheet.py for some hints).
but basically the beginning of the filename should repreresent a big feature of semgrep, e.g., dots_... metavar_... equivalence_... so here you probably want dots_field_chaining.js

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
regarding perf, it's not the only PR with some issues, so I don't think it's due to this PR; there's a little more job done, but not a lot.

regarding semgrep-rules, the best is to run make tests in semgrep-rules on a modified branch, with an updated semgrep-core in your PATH (do make install), to check that everything works fine.

**Comment by aryx** [comment]
It's a known problem: https://github.com/returntocorp/semgrep/issues/3539
I actually created a thread on slack to try to make semgrep-rules a submodule in semgrep-core/tests/
so we can be more sure that everything works together.

File: semgrep-core/src/matching/Generic_vs_generic.ml [hunk 4]
```diff
@@ -126,15 +126,21 @@ let rec obj_and_method_calls_of_expr e =
   match e.G.e with
   | B.Call ({ e = B.DotAccess (e, tok, fld); _ }, args) ->
       let o, xs = obj_and_method_calls_of_expr e in
-      (o, (fld, tok, args) :: xs)
+      (o, (fld, tok, Some args) :: xs)
+  | B.DotAccess (e, tok, fld) ->
```
Comment (aryx) [inline semgrep-core/src/matching/Generic_vs_generic.ml:130]: nice, I wrote the original code, and I'm surprised the fix was so simple :) I imagined worse.

**Comment by nmote** [comment]
So in the short term, how should I approach the issue with `semgrep-rules`? I've updated the PR description with the test I conducted, but that won't stop our CI from turning red when this PR is merged.

**Comment by aryx** [comment]
I'll try to add semgrep-rules as a submodule (as well as the test infra to use the tests in there), 
so then you can merge develop, modify locally semgrep-rules with the required modifications, so we can review and get confident that all of this can work together.

**Comment by nmote** [comment]
I've updated the `semgrep-rules` submodule to my branch with the updated test, and `make test` passes in `semgrep-core` for me locally (it didn't before).

**Comment by aryx** [comment]
Great! What I would do is open a PR at the same time with this branch you have in semgrep-rules to the semgrep-rules repo (and reference this PR here to say it's ok, so we get also a review from seceng on whether this looks legitimate).

**Comment by aryx** [comment]
I'll do it.

**Pull request merged**